### PR TITLE
Copy Middleware

### LINKED ISSUE #46: copy middleware
We need to implement copy middleware!

### PR DESCRIPTION
**Pull request by corystone**
Copy (and post as copy) middleware. Includes a couple of other changes, removed automagic compression in the http client and removed the quotes from etags, since we don't use them anywhere.

fixes #46

### CODE REVIEW

File: proxyserver/middleware/context.go [hunk 1]
```diff
@@ -121,6 +122,21 @@ func GetProxyContext(r *http.Request) *ProxyContext {
...
 
+func (ctx *ProxyContext) NewSubRequest(request *http.Request, method string, url string, body io.Reader) (*http.Request, error) {
+	subRequest, err := http.NewRequest(method, url, body)
+	if err != nil {
+		ctx = GetProxyContext(request)
+		ctx.Logger.Error("Couldn't create http.Request", zap.Error(err))
+		return nil, err
+	}
+	copyKeys := []string{"X-Auth-Token", "X-Trans-Id"}
+	for _, key := range copyKeys {
```
Comment (corystone) [inline proxyserver/middleware/context.go]: Agreed.

File: proxyserver/middleware/copy.go [hunk 2]
```diff
@@ -0,0 +1,433 @@
...
+		if !excluded {
+			dest.Del(k)
+			for _, v1 := range v {
+				dest.Add(k, v1)
+			}
+		}
+	}
+}
+
+func TrimEtag(etag string) string {
```
Comment (corystone) [inline proxyserver/middleware/copy.go]: Oops!

**Comment by corystone** [comment]
Hmm, here's what I'm doing and the results:

cory@saio:~$ curl -i http://127.0.0.1:8080/v1/AUTH_test/test/foo -X COPY -I -H "X-Auth-Token: 84bbaf40-3f5c-e339-e8f1-a0952f724646" -H "Destination:test/foo25"
HTTP/1.1 201 Created
Content-Length: 0
Content-Type: text/html
Last-Modified: Thu, 18 May 2017 08:25:32 GMT
X-Copied-From: test/foo
X-Copied-From-Account: AUTH_test
X-Delete-At:
X-Object-Meta-Foo: this is foos meta
X-Object-Meta-Mtime: 1494627621.709692
X-Trans-Id: 14bfa5d9f5e0e2bb
Date: Thu, 18 May 2017 08:25:56 GMT


cory@saio:~$ curl -i http://127.0.0.1:8080/v1/AUTH_test/test/foo26 -X PUT -I -H "X-Auth-Token: 20525564-5b61-da37-80ec-fbc04d019e70" -H "X-Copy-From: test/foo"
HTTP/1.1 201 Created
Content-Length: 0
Content-Type: text/html
Last-Modified: Thu, 18 May 2017 08:28:01 GMT
X-Copied-From: test/foo
X-Copied-From-Account: AUTH_test
X-Delete-At:
X-Object-Meta-Foo: this is foos meta
X-Object-Meta-Mtime: 1494627621.709692
X-Trans-Id: 14bfa5fcc88c3cc4
Date: Thu, 18 May 2017 08:28:26 GMT

File: proxyserver/middleware/copy.go [hunk 3]
```diff
@@ -0,0 +1,433 @@
...
+}
+
+func getSrcContainerObjectName(request *http.Request) (string, string, error) {
+	return getHeaderContainerObjectName(request, "X-Copy-From")
+}
+
+func getHeaderContainerObjectName(request *http.Request, header string) (string, string, error) {
+	path, err := url.QueryUnescape(request.Header.Get(header))
+	if err != nil {
+		msg := fmt.Sprintf("Invalid %s", header)
```
Comment (corystone) [inline proxyserver/middleware/copy.go]: Sure, no problem.

File: proxyserver/middleware/copy.go [hunk 4]
```diff
@@ -0,0 +1,433 @@
...
+}
+
+func CopyMetaItems(dest, src http.Header) {
+	CopyItemsWithPrefix(dest, src, "X-Object-Meta-")
+	CopyItemsWithPrefix(dest, src, "X-Object-Sysmeta-")
+	CopyItemsWithPrefix(dest, src, "X-Object-Transient-Sysmeta-")
+	dest.Set("X-Delete-At", src.Get("X-Delete-At"))
+}
+
+func CopyItemsExclude(dest, src http.Header, exclude []string) {
```
Comment (corystone) [inline proxyserver/middleware/copy.go]: Why would you ever pass a string you didn't want excluded?

File: proxyserver/middleware/copy.go [hunk 5]
```diff
@@ -0,0 +1,433 @@
...
+	w.w.Close()
+}
+
+func NewPipeResponseWriter(writer *io.PipeWriter, done chan bool, ctx *ProxyContext) *PipeResponseWriter {
+	header := make(map[string][]string)
+	return &PipeResponseWriter{
+		w:      writer,
+		header: header,
+		done:   done,
+		ctx:    ctx,
```
Comment (corystone) [inline proxyserver/middleware/copy.go]: Ya, it's not really using it. Good catch.

File: proxyserver/middleware/copy.go (hunk 4 above)
Comment (dpgoetz) [inline proxyserver/middleware/copy.go]: i just meant you could just test for membership in the map instead of walking through the array

Comment (corystone) [inline proxyserver/middleware/copy.go]: Ah, right.

**Pull request merged**